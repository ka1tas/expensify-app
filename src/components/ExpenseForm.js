import React from 'react';
import moment  from "moment";
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css'

export default class ExpenseForm extends React.Component{

    constructor(props) {
        super();
        this.state ={
            id: props.expense ? props.expense.id:'',
            description: props.expense? props.expense.description:'',
            amount: props.expense? (props.expense.amount*100).toString():'',
            note: props.expense? props.expense.note:'',
            createdDate: props.expense? moment(props.expense.createdDate ) : moment(),
            calenderFocused:false,
            error: '',
        }
    }




    onDescriptionChange = (e)=>{
        const description = e.target.value;
        this.setState(()=>({ description }));
    }

    onNoteChange=(e)=>{
        const note = e.target.value;
        this.setState(()=>({ note }));
    }

    onAmountChange=(e)=>{
        const amount = e.target.value;
        if(amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({ amount }));
        }

    }

    onDateChange = (createdDate) => {
        if(createdDate){
            this.setState(() => ({ createdDate }));
        }

    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }));
    };

    //description: '' , amount: '' , createdDate: moment() , note:''

    onSubmitForm = (e )=>{
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({error: <p> Description or Amount is missing. </p>}));
        } else {
            this.setState(() => ({
                error: ''
            }));
        }

        this.props.onSubmit({
             description: this.state.description,
             amount: parseFloat(this.state.amount, 10) * 100,
             note: this.state.note,
             createdDate: this.state.createdDate.valueOf()
        })
        }




    render(){
        console.log(this.props.expense);
        return (
            <div>
                <br/>
                { (!!this.state.error) && this.state.error}
                <br/>
            <form onSubmit={this.onSubmitForm}>
                <input type={'text'} placeholder={'Description'}
                value={this.state.description} onChange={this.onDescriptionChange}
                />
                <input type={'number'} placeholder={'Amount'}
                       value={this.state.amount} onChange={this.onAmountChange}/>
                <br/>
                <SingleDatePicker
                    date={this.state.createdDate}
                    onDateChange={this.onDateChange}
                    focused={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                />
                    <br/>
                <textarea placeholder={'Add a note related the expense (optional)'}
                          value={this.state.note} onChange={this.onNoteChange}
                ></textarea>
                <button> {(!this.state.id)?'Add Expense' : 'Edit Expense'}</button>
            </form>

            </div>
        )
    }
}