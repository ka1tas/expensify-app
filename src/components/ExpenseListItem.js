import React from 'react';
import {Link } from "react-router-dom";

export  const ExpenseListItem = ({dispatch , id, description , amount , createdDate })=>(
    <div>
        <h2><Link to={`/edit/${id}`}>{description}</Link></h2>
        <p> Amount : {amount} - Created At : {createdDate}</p>

    </div>
);



export default ExpenseListItem;

