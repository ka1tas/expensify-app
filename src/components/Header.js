import React from 'react';
import {NavLink} from "react-router-dom";

    const Header = ()=>(
    <div>
        <header>
            <h1>Expensify.</h1>
            <div>
                <h3>
                <span>
                    <NavLink to={'/'} activeClassName={'is-active'} exact={true}>Home  &nbsp;&nbsp; </NavLink>
                    <NavLink to={'/create'} activeClassName={'is-active'}>Add Expense   &nbsp;&nbsp;</NavLink>
                    <NavLink to={'/help'} activeClassName={'is-active'}>Help</NavLink>

                </span>
                </h3>
            </div>
        </header>
         <hr />
        </div>
    );

    export default Header;
