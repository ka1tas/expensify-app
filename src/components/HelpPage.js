import React from 'react';

    const HelpPage = () => (
        <div>
            <h2> This is the Help page.</h2>
        </div>
    );

    export default HelpPage;