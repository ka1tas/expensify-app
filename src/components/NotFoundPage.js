import React from 'react';
import {Link} from "react-router-dom";

    const NotFoundPage = () => (
        <div>
            <h2> 404 ! Page not found page.</h2>
           <h2>  <Link to={'/'}> Go back to home. </Link> </h2>
        </div>
    );

export default NotFoundPage;