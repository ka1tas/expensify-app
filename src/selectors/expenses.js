import moment from "moment";

//get visible expenses
export default (expenses , {text , sortBy , startDate, endDate})=>{
    return expenses.filter((expense)=>{
        const  createdAt = moment(expense.createdDate);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAt , 'day') : true ;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAt , 'day'): true;

        const textMatch = (text==='')?true:(expense.description.toLowerCase().includes(text.toLowerCase())?true:false)
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b)=>{
        if(sortBy === 'date'){
            return a.createdDate < b.createdDate ? 1 : -1;
        }
        else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

