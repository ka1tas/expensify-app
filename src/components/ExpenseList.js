import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses';
import ExpenseListFilter from "./ExpenseListFilters";

export const ExpenseList = (props)=>(
    <div>

        {(props.expenses.length===0)?
            <h1> No expense found! </h1> :

            props.expenses.map((expense)=>{
            return  <ExpenseListItem key={expense.id} {...expense}/>
        })


        }


    </div>
);

const mapStatetoProps =(state)=>{
    return {
        expenses: selectExpenses(state.expenses , state.filters)
    };
};

export default  connect(mapStatetoProps)(ExpenseList);

