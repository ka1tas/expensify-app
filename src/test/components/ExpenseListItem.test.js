import React from 'react';
import  {shallow} from 'enzyme';
import { ExpenseListItem } from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";


test('should render with expense ',()=>{
    const wraper = shallow(<ExpenseListItem {...expenses[2]}/>)
    expect(wraper).toMatchSnapshot();
})