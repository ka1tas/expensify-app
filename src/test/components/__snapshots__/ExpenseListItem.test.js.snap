// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/01"
  >
    <h3>
      Pen
    </h3>
  </Link>
  <p>
    $1.00
    -
    January 1st, 1970
  </p>
</div>
`;
